docs/DeveloperGuide.md

```diff
@@ -154,86 +154,116 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
 
-These operations are exposed in the `Model` interface as `Model#commitAddressBook()`, `Model#undoAddressBook()` and `Model#redoAddressBook()` respectively.
+`UndoableCommand` adds an extra layer between the abstract `Command` class and the concrete commands that can be undone such as `AddCommand`, `DeleteCommand`. `UndoableCommand` makes the `Command#execute()` method concrete, while the child classes implements an `executeUndoableCommand()` method that preprocesses the Command before execution. Each child class knows how to undo/redo by itself.
 
-Given below is an example usage scenario and how the undo/redo mechanism behaves at each step.
+Commands that are not undoable have similar implementations as the code snippet below:
 
-Step 1. The user launches the application for the first time. The `VersionedAddressBook` will be initialized with the initial address book state, and the `currentStatePointer` pointing to that single address book state.
+    public class ListCommand extends Command {
+        @Override
+        public CommandResult execute() {
+            //..list logic
+        }
+    }
 
-![UndoRedoState0](images/UndoRedoState0.png)
 
-Step 2. The user executes `delete 5` command to delete the 5th person in the address book. The `delete` command calls `Model#commitAddressBook()`, causing the modified state of the address book after the `delete 5` command executes to be saved in the `addressBookStateList`, and the `currentStatePointer` is shifted to the newly inserted address book state.
+With `UndoableCommand`, the commands that are undoable are implemented similarly to the code snippet below:
 
-![UndoRedoState1](images/UndoRedoState1.png)
+    public abstract class UndoableCommand extends Command {
+        protected abstract CommandResult executeUndoableCommand();
+        @Override
+        public CommandResult execute() {
+            return executeUndoableCommand();
+        }
+        protected abstract void undo;
+    }
+    
+    public class DeleteCommand extends UndoableCommand {
+        @Override
+        protected CommandResult executeUndoableCommand() {
+            //...delete logic
+        }
+        @Override
+        protected void undo {
+            //...undo logic
+        }
+    }
 
-Step 3. The user executes `add n/David …​` to add a new person. The `add` command also calls `Model#commitAddressBook()`, causing another modified address book state to be saved into the `addressBookStateList`.
+Step 1. Suppose that a user has just launched our application. The `undoRedoStack` will be empty at the beginning.
 
-![UndoRedoState2](images/UndoRedoState2.png)
+Step 2. The user then executes an `UndoableCommand`. The command would modify the existing `Model` as per the definition of an `UndoableCommand`. 
+The specific called `UndoableCommand` will process the command during the method call `UndoableCommand#executeUndoableCommand` so that it can undo itself.
```

Comment:
```suggestion
The specific `UndoableCommand` called will process the command during the method call `UndoableCommand#executeUndoableCommand` so that it can undo itself.
```